### PR TITLE
[MNG-4660] Add integration test to show that --resume-from works

### PR DESCRIPTION
**Pull request by mthmulders**
Will offer this one for merging once the bug has indeed been fixed, and the `// TODO` regarding the version range has been resolved.

### CODE REVIEW

File: core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java [hunk 1]
```diff
@@ -231,6 +231,7 @@ public static Test suite()
         suite.addTestSuite( MavenITmng4679SnapshotUpdateInPluginTest.class );
         suite.addTestSuite( MavenITmng4677DisabledPluginConfigInheritanceTest.class );
         suite.addTestSuite( MavenITmng4666CoreRealmImportTest.class );
+        suite.addTestSuite( MavenITmng4660ResumeFromTest.class );
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: In general this is in order of implementation, not in order of reporting (also helps with fast feedback)

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        finally
+        {
+            verifier1.resetStreams();
+        }
+
+        final Verifier verifier2 = newVerifier( testDir.getAbsolutePath() );
+        verifier2.setAutoclean( false );
+        verifier2.addCliOption( "--resume-from" );
+        verifier2.addCliOption( ":module-b" );
+        verifier2.executeGoal( "compile" ); // to prevent the unit test from failing (again)
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: This looks like a mixup of 2 separate issues:
1. you want to continue and reuse the artifacts that have already been build. In that case the goal should be `verify`.
2. you want to compile module-b, but since Maven hasn't attached its artifacts to the reactor, you need to recompile module-a too. In that case I would add `verifier2.addCliOption( "--also-make" );`

Both are interesting and don't work right now. 1 is harder (and matches the description of the issue), 2 should be reasonably easier and already a huge improvement, but requires a separate issue.

Comment (rfscholte) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: BTW I do like this usecase: ignore the testresult of an early module for now, continue with its artifact to see the result for the rest of the project.

File: core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java (hunk 1 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: [Line 723](https://github.com/apache/maven-integration-testing/blob/e8f24ba23d9a8d97a2e1e8c9920213ffb6ebe772/core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java#L723) suggests something else...:

>  Add tests in reverse alpha order above.

Comment (rfscholte) [reply core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: I see, and https://github.com/apache/maven-integration-testing/blob/e8f24ba23d9a8d97a2e1e8c9920213ffb6ebe772/core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java#L88-L93 seems contradictionary. This text must be from the period we didn't start with MNG-nnn yet. However, the intention is right: newer tests are more likely to fail, so start with those.
I'll rewrite it...

Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: OK, I'll await the outcome of that and then update the place of this new test in the suite.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java (hunk 2 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: We do option 2, but instead of recompiling module-a, we want to reuse the outcome of the previous build, which did include module-a. That's why we have `--resume-from :module-b` and _autoclean_ set to off. The test that resides in module-b would fail again and that's why we went for `compile`.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * This is a test case for <a href="https://issues.apache.org/jira/browse/MNG-4660">MNG-4660</a>.
+ *
+ * @author Maarten Mulders
+ * @author Martin Kanters
+ */
+public class MavenITmng4660ResumeFromTest extends AbstractMavenIntegrationTestCase {
+    public MavenITmng4660ResumeFromTest()
+    {
+        super( "[3.7.0-SNAPSHOT,)" );
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: This should be 3.7.0, the  `AbstractMavenIntegrationTestCase` understands that you need to test this a SNAPSHOT.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java [hunk 4]
```diff
@@ -61,7 +60,7 @@ public void testitShouldResolveTheDependencies()
         
         verifier.addCliOption( "-Drevision=1.2" );
         verifier.setLogFileName( "install-log.txt" );
-        verifier.executeGoals( Arrays.asList( "clean", "install" ) );
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: Please don't change other integration tests, only the version-range if the behavior changes (or add code)
I'm pretty sure you want to test the `install` phase when working with CI-friendly versions, to ensure that still works.

Comment (MartinKanters) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: Right, I'm not sure if this was an intended change. I've reverted it.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java (hunk 3 above)
Comment (MartinKanters) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: Interesting! Fixed.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java (hunk 4 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: It was an intended change, we figured that `clean install` was no longer needed. But I guess it'll still work 😄

File: core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java (hunk 1 above)
Comment (MartinKanters) [reply core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: I've put the test at the top of the list. Do you want us to change the javadoc (pointed out by Maarten above) or do you want to do that separately?

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+        verifier1.verifyErrorFreeLog();
+        verifier1.resetStreams();
+
+
+        // 2. Create a properties file with some content and compile only that module (module A).
+        final Verifier verifier2 = newVerifier( testDir.getAbsolutePath() );
+
+        final Path resourcesDirectory = Files.createDirectories( Paths.get( testDir.toString(), "module-a", "src", "main", "resources" ) );
+        System.err.println( "Created directory " + resourcesDirectory.toAbsolutePath().toString() );
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java]: System.x.println don't belong in tests

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        final Path fileToWrite = resourcesDirectory.resolve( "example.properties" );
+        FileUtils.fileWrite( fileToWrite.toString(), "x=42" );
+        System.err.println( "Wrote file " + fileToWrite.toAbsolutePath().toString() );
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java]: Same here

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        super( "[3.7.0,)" );
+    }
+
+    /**
+     * Test that the --resume-from flag resolves dependencies inside the same Maven project
+     * without having them installed first.
+     * This test case uses the target/classes folder of module-a.
+     */
+    public void testShouldResolveUnpackagedArtifactFromEarlierBuild() throws Exception
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: method name is a bit confusing. 'Unpackaged' might suggest the artifact is unpackaged in the process. Here you want to use the outputDirectory.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java (hunk 5 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java]: Oops, removed.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java (hunk 6 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660OutdatedPackagedArtifact.java]: Oops, removed.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java [hunk 8]
```diff
@@ -19,7 +19,6 @@
  * under the License.
  */
 
-import org.apache.maven.it.Verifier;
```
Comment (rfscholte) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: unrelated change

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java (hunk 7 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng4660ResumeFromTest.java]: Agreed, updated method name as well as Javadoc.

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java (hunk 8 above)
Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: True. We were investigating that test case and probably the IDE thought it'd be good to cleanup unused imports. Prefer to have it reverted? I think it won't hurt either....

Comment (rfscholte) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: It will look very weird if we go through the commit logs and see the list of files touched by this commit. It scares me if I see MNG6090 between MNG4660

Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6090CIFriendlyTest.java]: Don't be scared, it's been reverted already.

File: core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java (hunk 1 above)
Comment (rfscholte) [reply core-it-suite/src/test/java/org/apache/maven/it/IntegrationTestSuite.java]: I will update the javadoc. I've found several places that require an update
